### DISCUSSION TITLE
loom virtual threads

### DISCUSSION THREAD

**Discussion by pnf** [async-profiler/async-profiler#846] [General]
I was surprised to discover that flight recorder seems to handle virtual threads reasonably well. For example if you have
```
     def burnCpu(): Double =  ... //expensive
     def bar() = { 
          // call expensive method twice, separated by some suspendable call
          val x = burnCpu();
          Thread.yield()    
          x + burnCpu() 
     }
     def foo() = {
         // call bar twice, separated by some suspendable call
         val x = bar()
         Thread.yield()
         x + bar() 
     }
     // Launch a whole bunch of virtual threads calling foo
     val tasks = (1 to n).map(i => scope.fork(() => foo()))
     scope.join()
```
then jfr shows foo -> bar -> burn as one would hope:
<img width="395" alt="image" src="https://github.com/async-profiler/async-profiler/assets/1265645/22d90540-7dee-4a0b-8045-661ec911aa18">
while jfr generated from async-profiler shows a mixture of stacks, some omitting the call to foo altogether,
<img width="801" alt="image" src="https://github.com/async-profiler/async-profiler/assets/1265645/84bbbb88-9df3-4726-89ed-29c19c14e10a">
when the virtual thread happens to get picked up by a carrier with nothing on its stack when continuing after the yield in bar.

The async-profiler results are exactly what I would expect; the surprise was that flight recorder somehow is restoring the full stack after the continuation.  Moreover, it manages to do this even if I throw in a few layers of explicit virtual thread tasks:
```
  def innerTask(i: Int, j: Int)(): Double = {
    foo()
  }
  def outerTask(i: Int)(): Double = {
    val scope = new StructuredTaskScope.ShutdownOnFailure()
    val tasks = (1 to nInner).map(j => scope.fork(innerTask(i, j) _))
    scope.join()
    val ret = tasks.map(_.get()).sum
    ret
  }
  val scope = new StructuredTaskScope.ShutdownOnFailure()
  val tasks = (1 to nOuter).map(i => scope.fork(outerTask(i) _))
  scope.join()
```
<img width="413" alt="image" src="https://github.com/async-profiler/async-profiler/assets/1265645/a6d4908d-ebc8-4cee-aade-e9cefa5d40ee">

So first, how is flight recorder doing this (something other that ASGCT?), and, second, is it plausible that async-profiler could do the same thing, and still be able to sample native stacks?

**Comment by franz1981** [comment]
I would double check what jfr does but...how many samples jfr seems to collect? There is some chance that it doesn't restore anything but just drop silently continuations (resume/blocked) stacks because both involves JIT to physically adjust the frame pointers, last time I have checked

**Comment by franz1981** [comment]
This is the entry point for what jfr does (should follow the pointers) https://github.com/openjdk/jdk/blob/jdk-21%2B35/src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp#L263 for JDK 21

**Reply by franz1981** [reply]
And found something at https://github.com/openjdk/jdk/blob/jdk-21%2B35/src/hotspot/cpu/x86/frame_x86.cpp#L158 which is related to adding virtual thread support

**Reply by franz1981** [reply]
But in theory they should both (AGCT and jfr) hit the same methods, which should include the v thread support.

**Comment by parttimenerd** [comment]
Just a pointer, as I hadn't time to properly look into it, but I vaguely remember that JFR tries to reconstitute some continuations:

```java
bool JfrThreadSampleClosure::sample_thread_in_java(JavaThread* thread, JfrStackFrame* frames, u4 max_frames) {
  // Process the oops in the thread head before calling into code that wants to
  // stack walk over Loom continuations. The stack walking code will otherwise
  // skip frames in stack chunks on the Java heap.
  StackWatermarkSet::start_processing(thread, StackWatermarkKind::gc);

  OSThreadSampler sampler(thread, *this, frames, max_frames);
  sampler.take_sample();
```
(https://github.com/openjdk/jdk/blob/d0761c19d1ddafbcb5ea97334335462e716de250/src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp#L283)

If I remember correctly, this code isn't present in ASGCT.

**Reply by franz1981** [reply]
thanks @parttimenerd I was going in to nerd snipe you anyway ihih

**Comment by parttimenerd** [comment]
> second, is it plausible that async-profiler could do the same thing, and still be able to sample native stacks?

No. Wait for my JEP :)

**Comment by apangin** [comment]
Async-profiler collects stack traces of carrier threads. In order to get stack traces of virtual threads, one would need to safely decode internal on-heap structures. JFR, being a part of the JVM, is capable of doing so. It's further easier for the JFR to parse virtual stacks, because it performs sampling in one dedicated thread. This solution, although does not scale well, makes it easy to coordinate work with the rest of the JVM, e.g., GC.
Async-profiler works on the native threads level. Reconstruction of virtual stack traces is not its goal at this point.

**Comment by pnf** [comment]
Unfortunately, this makes async-profiler essentially incompatible with loom.  If the problem was limited to reconstruction of the chain of deliberately launched structured concurrency scopes, that would probably not be an issue for most applications, but if any potentially yielding method can be yanked onto a random carrier thread without any stack history, the flame graphs will be very hard to interpret.
